src/abbreviation/rewriter/AbbreviationRewriter.ts

```diff
@@ -0,0 +1,256 @@
+import * as vscode from 'vscode';
+import { commands, Disposable, TextEditor, window, workspace } from 'vscode';
+import { assert } from '../../utils/assert';
+import { AbbreviationProvider } from '../AbbreviationProvider';
+import { AbbreviationConfig } from '../config';
+import { Range } from './Range';
+import { TrackedAbbreviation } from './TrackedAbbreviation';
+
+/**
+ * Tracks abbreviations in a given text editor and replaces them when dynamically.
+ */
+export class AbbreviationRewriter {
+	private readonly disposables = new Array<Disposable>();
+	/**
+	 * All tracked abbreviations are disjoint.
+	 */
+	private readonly trackedAbbreviations = new Set<TrackedAbbreviation>();
+	private readonly decorationType = window.createTextEditorDecorationType({
+		textDecoration: 'underline',
+	});
+
+	private dontTrackNewAbbr = false;
+
+	constructor(
+		private readonly config: AbbreviationConfig,
+		private readonly abbreviationProvider: AbbreviationProvider,
+		private readonly textEditor: TextEditor
+	) {
+		this.disposables.push(this.decorationType);
+
+		this.disposables.push(
+			workspace.onDidChangeTextDocument((e) => {
+				if (e.document !== this.textEditor.document) {
+					return;
+				}
+
+				const changes = e.contentChanges.slice(0);
+				// We need to process the changes at the bottom first.
+				// Otherwise, changes at the top will move spans at the bottom down.
+				changes.sort((c1, c2) => c2.rangeOffset - c1.rangeOffset);
+
+				for (const c of changes) {
+					const range = new Range(c.rangeOffset, c.rangeLength);
+					this.processChange(range, c.text);
+				}
+
+				this.updateState();
+
+				// Replace any tracked abbreviation that is either finished or unique.
+				void this.forceReplace(
+					[...this.trackedAbbreviations].filter(
+						(abbr) =>
+							abbr.finished ||
+							(this.config.eagerReplacementEnabled.get() &&
+								abbr.isAbbreviationUniqueAndComplete)
+					)
+				);
+			})
+		);
+		this.disposables.push(
+			window.onDidChangeTextEditorSelection((e) => {
+				if (e.textEditor !== this.textEditor) {
+					return;
+				}
+
+				// Replace any tracked abbreviation that lost selection.
+				void this.forceReplace(
+					[...this.trackedAbbreviations].filter(
+						(abbr) =>
+							!e.selections.some((s) =>
+								abbr.range.containsRange(
+									fromVsCodeRange(
+										s,
+										e.textEditor.document
+									).withLength(0)
+								)
+							)
+					)
+				);
+			})
+		);
+
+		this.disposables.push(
+			commands.registerTextEditorCommand('lean.input.convert', async () =>
+				this.forceReplace([...this.trackedAbbreviations])
+			)
+		);
+	}
+
+	private async forceReplace(
+		abbreviations: TrackedAbbreviation[]
+	): Promise<void> {
+		if (abbreviations.length === 0) {
+			return;
+		}
+		for (const a of abbreviations) {
+			this.trackedAbbreviations.delete(a);
+		}
+
+		// Wait for VS Code to trigger `onDidChangeTextEditorSelection`
+		await waitForNextTick();
```

Comment:
My understanding is that this timeout is non-deterministic, and is the cause of the "sometimes the replacement doesn't happen" bug. It happens more when lean is busy.

Reply:
> "sometimes the replacement doesn't happen"

This happens if the user changes the text buffer before the replacment is committed. This is unavoidable, it can only be made less likely to happen (I only learned recently about this, sorry for not having taken this into account when I wrote that code).

Internally, VS Code has a version number per text document that is incremented on every edit. Edits are sent asynchronously to the extensions and if an extension issues an (async) text modification it is only applied if the version of the target document has not changed in the meantime.

`waitForNextTick` makes it more likely that a user can change the text buffer before the extension can commit the text modification.

However, `waitForNextTick` is also required to position the cursor correctly after the replacment has been committed.

Your fix is a good workaround, but maybe the extension can actually try a second time to perform this replacement. Sadly (afaik) VS Code does not allow transactional updates that involve text and cursor modification.